### PR TITLE
MGMT-10405 UI changes for schedulable masters

### PR DESCRIPTION
**Pull request by celdrake**
https://issues.redhat.com/browse/MGMT-10405

The UI now determines if the Switch for "schedulable_masters" is enabled or disabled based on the API's new field `schedulable_masters_forced_true`. This now allows the user to have schedulable_masters even when there are 5+ hosts.

For existing clusters, we may not receive the new field, in which case the UI keeps the existing logic (to be removed in a few weeks when all clusters receive the new field)

### CODE REVIEW

File: src/common/api/swagger.yaml [hunk 1]
```diff
@@ -1002,6 +1002,11 @@ paths:
           type: string
           format: uuid
           required: true
+        - in: query
+          name: timestamp
+          type: integer
+          description: The time on the host as seconds since the Unix epoch.
+          required: false
```
Comment (jgyselov) [inline src/common/api/swagger.yaml:1009]: Please make sure you are only updating the parts of `swagger.yaml` that are relevant to this issue.

File: src/common/api/types.ts [hunk 2]
```diff
@@ -220,12 +226,6 @@ export interface Cluster {
    * {"networking":{"networkType": "OVNKubernetes"},"fips":true}
    */
   installConfigOverrides?: string;
-  /**
-   * Json formatted string containing the user overrides for the initial ignition config
-   * example:
-   * {"ignition": {"version": "3.1.0"}, "storage": {"files": [{"path": "/tmp/example", "contents": {"source": "data:text/plain;base64,aGVscGltdHJhcHBlZGluYXN3YWdnZXJzcGVj"}}]}}
-   */
-  ignitionConfigOverrides?: string;
```
Comment (jgyselov) [inline src/common/api/types.ts:228]: Same as `swagger.yaml`. We only update the parts relevant to the specific issue.

File: src/common/selectors/clusterSelectors.ts [hunk 3]
```diff
@@ -89,3 +89,27 @@ export const selectIpv6HostPrefix = ({ clusterNetworks }: Pick<Cluster, 'cluster
...
+};
+
+export const selectSchedulableMasters = (
+  cluster: Pick<Cluster, 'schedulableMasters' | 'schedulableMastersForcedTrue'>,
+): boolean => {
+  if (selectMastersMustRunWorkloads(cluster)) {
+    return true;
+  }
+  return cluster.schedulableMasters || false;
+};
```
Comment (jgyselov) [inline src/common/selectors/clusterSelectors.ts:115]: Is it smart to use `Pick<>` if we aren't actually picking anything?

File: src/ocm/components/clusterWizard/HostDiscovery.tsx [hunk 4]
```diff
@@ -4,10 +4,10 @@ import { Formik, FormikConfig, useFormikContext } from 'formik';
 import {
   Cluster,
   V2ClusterUpdateParams,
-  getFormikErrorFields,
   ClusterWizardStep,
   useAlerts,
   getHostDiscoveryInitialValues,
+  getFormikErrorFields,
```
Comment (jgyselov) [inline src/ocm/components/clusterWizard/HostDiscovery.tsx:10]: Necessary change? :stuck_out_tongue:

**Review by jgyselov — CHANGES_REQUESTED** [review]

File: src/common/api/swagger.yaml (hunk 1 above)
Comment (celdrake) [reply src/common/api/swagger.yaml]: True, thanks for pointing that out

File: src/ocm/components/clusterWizard/HostDiscovery.tsx (hunk 4 above)
Comment (celdrake) [reply src/ocm/components/clusterWizard/HostDiscovery.tsx]: Oops! and it's gone :smile:

File: src/common/selectors/clusterSelectors.ts (hunk 3 above)
Comment (celdrake) [reply src/common/selectors/clusterSelectors.ts]: I see what you mean. Yes, doesn't make a lot of sense in this case.
Fixed it!

**Review by jgyselov — APPROVED** [review]

**Review by jgyselov — DISMISSED** [review]

**Review by jgyselov — APPROVED** [review]

File: src/ocm/components/clusterConfiguration/HostInventory.tsx [hunk 5]
```diff
@@ -68,23 +68,26 @@ const SchedulableMastersLabel: React.FC = () => (
...
+const HostInventory = ({ cluster }: { cluster: Cluster }) => {
   const { isPlatformIntegrationSupported } = useClusterSupportedPlatforms(cluster.id);
   const isPlatformIntegrationFeatureEnabled = useFeature(
     'ASSISTED_INSTALLER_PLATFORM_INTEGRATION_FEATURE',
   );
   const isOpenshiftClusterStorageEnabled = useFeature('ASSISTED_INSTALLER_OCS_FEATURE');
   const isContainerNativeVirtualizationEnabled = useFeature('ASSISTED_INSTALLER_CNV_FEATURE');
   const isSNOCluster = isSNO(cluster);
-  const isSchedulableMastersEnabled = !schedulableMastersAlwaysOn(cluster);
+  const mastersMustRunWorkload = selectMastersMustRunWorkloads(cluster);
```
Comment (jkilzi) [inline src/ocm/components/clusterConfiguration/HostInventory.tsx:83]:
```suggestion
  const mastersMustRunWorkloads = selectMastersMustRunWorkloads(cluster);
```

File: src/ocm/components/clusterConfiguration/HostInventory.tsx [hunk 6]
```diff
@@ -68,23 +68,26 @@ const SchedulableMastersLabel: React.FC = () => (
...
+
   const { setFieldValue } = useFormikContext<HostDiscoveryValues>();
+
   React.useEffect(() => {
-    setFieldValue('schedulableMasters', getSchedulableMasters(cluster));
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [isSchedulableMastersEnabled]); //just when changes from disabled to enabled, shouldn't respond to continous polling otherwise
+    setFieldValue('schedulableMasters', selectSchedulableMasters(cluster));
+  }, [mastersMustRunWorkload]); // Schedulable masters need to be recalculated only when forced status changes
```
Comment (jkilzi) [inline src/ocm/components/clusterConfiguration/HostInventory.tsx:89]:
```suggestion
  }, [mastersMustRunWorkloads]); // Schedulable masters need to be recalculated only when forced status changes
```

File: src/ocm/components/clusterConfiguration/HostInventory.tsx [hunk 7]
```diff
@@ -139,10 +142,10 @@ const HostInventory: React.FC<{ cluster: Cluster }> = ({ cluster }) => {
       <StackItem>
         <SwitchField
           tooltipProps={{
-            hidden: isSchedulableMastersEnabled,
+            hidden: !mastersMustRunWorkload,
             content: schedulableMastersTooltip,
           }}
-          isDisabled={!isSchedulableMastersEnabled}
+          isDisabled={mastersMustRunWorkload}
```
Comment (jkilzi) [inline src/ocm/components/clusterConfiguration/HostInventory.tsx:148]:
```suggestion
          isDisabled={mastersMustRunWorkloads}
```

**Review by jkilzi — DISMISSED** [review]

**Review by jkilzi — APPROVED** [review]

**Pull request merged**